_posts/2021-01-27-#20833.md

```diff
@@ -0,0 +1,164 @@
+  limited to one transaction at a time, which means users cannot test
+  transactions that depend on outputs that the mempool has not seen yet. The main
+  reason for this limitation is that our validation code only uses the coins
+  available in the current chain state and mempool (and not anything else), to
+  validate transaction inputs.
+
+- Another reason package mempool accept may be of interest is to improve fee
+  bumping.  Currently, Child Pays for Parent does not work for transactions
+  with a feerate below the mempool minimum fee, which can be dangerous for
+  recipients of time-sensitive transactions.
+  [#20833](https://github.com/bitcoin/bitcoin/pull/20833) is a first step towards
+  package validation and hopefully, eventually, [Package
+  Relay](https://bitcoinops.org/en/topics/package-relay/).  We have discussed
+  Package Relay in a previous review club,
+  [#16401](https://bitcoincore.reviews/16401). 
+
+- You can read more about Package Mempool Acceptance
+  [here](https://brink.dev/blog/2021/01/21/fellowship-project-package-accept/).
+
+## Notes
+
+- [#20833](https://github.com/bitcoin/bitcoin/pull/20833) enables dry runs of
+  package validation.  It does this by having `CCoinsViewMemPool` keep track of
+  the Coins created and spent by transactions in the package.
+
+- The
+  [`MemPoolAccept`](https://github.com/bitcoin/bitcoin/blob/7f653c3b/src/validation.cpp#L451)
+  class represents a "validation session" with the current mempool and chain
+  state; it holds `cs_main` and `pool.cs` mutexes throughout validation.  Its job
+  is to figure out whether a transaction (received from peers or clients like
+  RPC) is a valid candidate for submission to the mempool.  To do so, it applies
+  consensus and
+  [policy](https://github.com/bitcoin/bitcoin/tree/c831e105/src/policy) rules.
+  Importantly, if a transaction is invalid, `MemPoolAccept` is designed such that
+  it fails as quickly as possible to mitigate DoS risks.
+
+- Each `MemPoolAccept` instance creates a Coins View from the current chain tip
+  and mempool.  `CCoinsView` is a class with a hierarchical structure (see
+  diagram below).
+
+```
+             +--------------------+
+             |                    |
+             |     CCoinsView     |
+             |                    |
+             +----------+---------+
+                        ^
+                        |
+             +----------+---------+
+             |                    |
+             |  CCoinsViewBacked  |
+             |                    |
+             +------+------+------+
+                    ^      ^
+           +--------+      +--------+
+           |                        |
++----------+---------+  +-----------+---------+
+|                    |  |                     |
+|  CCoinsViewCache   |  |  CCoinsViewMemPool  |
+|                    |  |                     |
++--------------------+  +---------------------+
```

Comment:
Nice diagram!

Perhaps it'd be even clearer if we used this:

![image](https://user-images.githubusercontent.com/1063656/105534906-146f3d00-5ce6-11eb-9b05-734636d4a969.png)

(With @jamesob's permission)

Also maybe link to http://localhost:4000/17487, which was a good review club meeting on the coins classes.

Reply:
Are you saying we have jamesob's permission, or I should ask?

Reply:
oo also, we don't have `AcceptToMemoryPoolWorker` anymore btw

Reply:
Ah ok, maybe just use the existing diagram.

> Are you saying we have jamesob's permission, or I should ask?

That was unclear! I meant we should ask, but that's moot now :)